src/security/context.cpp

```diff
@@ -643,16 +694,197 @@ class PolFinalWafCtx : public PolTaskCtx<PolFinalWafCtx> {
 
   void complete() noexcept {
     ngx_log_debug0(NGX_LOG_DEBUG_HTTP, req_.connection->log, 0,
-                   "completion handler of waf task");
+                   "completion of final waf task");
   }
 
   friend PolTaskCtx;
 };
 
+ngx_int_t Context::do_request_body_filter(ngx_http_request_t &request,
+                                          ngx_chain_t *in, dd::Span &span) {
+  auto st = stage_->load(std::memory_order_acquire);
+  ngx_log_debug3(
+      NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+      "request body filter %s in chain. Current data: %lu, Stage: %d",
+      in ? "with" : "without", filter_ctx_.out_total, st);
+
+  if (st == stage::AFTER_BEGIN_WAF) {
+    ngx_log_debug1(NGX_LOG_DEBUG_HTTP, request.connection->log, 0,
+                   "first filter call, req refcount=%d", request.main->count);
+    // buffer the request body during reading
+    // https://github.com/nginx/nginx/commit/67d160bf25e02ba6679bb6c3b9cbdfeb29b759de
```

Comment:
The lowest version we build for 1.22.0. Are we supposed to support older versions? Because I can either disable this feature for nginx < 1.21.4 or just note in the README

Reply:
> Are we supposed to support older versions?
No, we officially support only the last 2 years of NGINX. You'r good, please juste add a note in the README.